[verifier] Add option to validate array(float) via error margin.

spershin opened the pull request:
## Description
Basically allow verifier to validate array(float) and array(double) columns in the same way as we validate float and double columns. Later we can extend this to the maps as well.


## Motivation and Context
We are planning to use this when comparing Presto Native with Presto Java to reduce the noise from array(double) columns. This will also strengthen the Presto verification as non-deterministic array(double) column could hide a valid correctness issue in another column by marking query SKIPPED in the verifier.

## Test Plan
Added new unit test and updated existing one.
Ran verifier on a query producing non-deterministic array(double) (artificial query).
Result is successful: the query passes correctness in the new verifier, while getting skipped in the old one.
Successfully run real use case verifier suite on the new verifier.

```
== NO RELEASE NOTE ==
```

spershin commented:
Submitted PR a bit too early - the validation part has not been implemented yet. But the PR is good to see where it is all going.

spershin commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:
```diff
@@ -23,12 +23,27 @@ public class ArrayColumnChecksum
-    public ArrayColumnChecksum(@Nullable Object checksum, @Nullable Object cardinalityChecksum, long cardinalitySum)
+    public ArrayColumnChecksum(
+            @Nullable Object checksum,
+            @Nullable Object sum,
```
@elharo 
I'm not sue, I was following the existing code from FloatingPointColumnValidator.
As I can see classes have long for counts and Object for everything else, why - not sure. :)

elharo replied:
Let's watch out for Chesterton's Fence. Before proceeding we should figure out why checksum is an Object. Presumably there's a reason but maybe that reason applies to the new cardinality checksum and maybe it doesn't. It's also possible checksum shouldn't have been an Object in the first place, in which case we don't have to fix this now, but let's not repeat the mistake.  Any idea who could shed some light on this?

rschlussel replied:
I can't see a good reason why sum is an Object in FloatingPointColumnChecksum.  the only places it's used, it's blindly cast to a double https://github.com/prestodb/presto/blob/master/presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/FloatingPointColumnValidator.java#L145-L146. For arrays, the sum is not necessarily a double, and I think it would be better not to assume it's a double here. (i mean you've only implemented it for double type, but that's more incidental.  You should be able to do sum validation on arrays of other numeric types)

Regarding why Objects are used for checksums throughout, I'm not sure. I assume it was most convenient, but I also suspect the code could be improved to use more specific types. In case it's helpful for gleaning more, the object usage seems to have been introduced here https://github.com/prestodb/presto/pull/12489/commits/04d45ea146cc858142e9fbd460bfe5e3ded82856 and here: https://github.com/prestodb/presto/pull/14506/commits/7c481036fb85f32fb7ff4c9fc0762f5df6389946 if that helps at all.

spershin replied:
@elharo 

> Let's watch out for Chesterton's Fence. 

If we follow that rule you mentioned - we should not deviate of what already is in there. I suggest approach of 'as is'  and then follow up if we decide to change this. I would like this PR to stay focused on its task, if possible.

@rschlussel 

> Some additional thoughts - since we could imagine wanting to add special validation for other types.

Yes, we would likely go that way. Maps of doubles is the next candidate, especially for values part. I was thinking about it, however wanted to stay focused on just arrays now to unblock NGA use case quickly. We can adjust few things to fit Maps more as we go and introduce more unconventional support.

> Finally, I think we have the same problem you're solving here for map types, where it does checksums on the keys and values.

Not only that, but the structural types. like rows, but not sure if we want to really go there. Let's fo one step at a time? Harder to break existing stuff that way.

spershin replied:
@elharo @rschlussel 
I believe objects are being used, because the aggregation can return NULL when there are no rows.
I'm currently debugging exactly such case.
To handle these nulls we use Object rather than a direct type.

rschlussel replied:
Interesting. but then why not use a Double type (which can be null since it's an object) rather than a double primitive type?

spershin replied:
Good question.
We didn't think about in the beginning?

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -16,56 +16,86 @@
-import com.facebook.presto.sql.tree.QualifiedName;
-import com.facebook.presto.sql.tree.SingleColumn;
-import com.facebook.presto.sql.tree.TryExpression;
+import com.facebook.presto.sql.tree.*;
```
don't use * imports

spershin replied:
These are added automatically by the IJ, I think, will remove, but got to keep an eye on it.

spershin commented:
The latest update includes:
1. Fixed failing tests.
2. Refactored validate() from FloatingPointColumnValidator class to FloatingPointColumnChecksum.
3. Fixed query generation for array(double) validator (infinity issue).
4. Added TestChecksumValidator.testFloatingPointArray() unit test.
5. Added support of array(double) and array(real) to TestChecksumValidator.testChecksumQuery() unit test.
6. Improved TestChecksumValidator.testChecksumQuery() to tell which line does not match.

spershin commented:
Fixed Checkstyle errors.

spershin commented:
More Checkstyle errors...

spershin commented:
Aaaand Checkstyle again...

spershin commented:
Fixed the checksum query.

kewang1024 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -16,82 +16,205 @@
+    }
+
+    @Override
+    public List<ColumnMatchResult<ArrayColumnChecksum>> validate(Column column, ChecksumResult controlResult, ChecksumResult testResult)
```
NIT: Can we put this function to its original place? (before function generateArrayChecksum). we should put public functions ahead then private function

spershin replied:
I placed them like this of better see the logic of the code:
public generateChecksumColumns() uses the following generateArrayChecksum(), generateInfinityLambdaExpression() and generateLambdaExpression().

public validate() uses the following toColumnChecksum().

I can bring them back, no issue.

spershin replied:
Will be addressed in the next push.

kewang1024 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -16,82 +16,205 @@
+                testResult.getRowCount(),
+                controlResult.getRowCount());
+
+        boolean useFloatingPointPath = useFloatingPointPath(column);
+
+        ArrayColumnChecksum controlChecksum = toColumnChecksum(column, controlResult, useFloatingPointPath);
+        ArrayColumnChecksum testChecksum = toColumnChecksum(column, testResult, useFloatingPointPath);
+
+        // Not floating point case (we have '$checksum' column).
+        if (!useFloatingPointPath) {
```
We don't need `useFloatingPointPath ` here, we can infer this information from ArrayColumnChecksum structure

spershin replied:
How so?

kewang1024 replied:
if `ArrayColumnChecksum `'s checksum is not null, then it's regular ArrayColumnChecksum
If it is null, then it's the float point check

spershin replied:
> if `ArrayColumnChecksum `'s checksum is not null, then it's regular ArrayColumnChecksum If it is null, then it's the float point check

@kewang1024 

That was my 1st choice, but it all breaks on the corner case when we get checksums on empty tables (no rows) and then a bunch of checksum columns have nulls.

I realized after that if we are triggering floating point path checking the array element type - we should be consitent and always check using that approach. :)

kewang1024 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ChecksumResult.java:
```diff
@@ -75,6 +75,11 @@ public Map<String, Object> getChecksums()
         return checksums;
     }
 
+    public boolean hasChecksum(String columnName)
```
Looks like not used in this PR?

spershin replied:
Ah, yes, I used 1st and then had to go different path because of the NULLs corner case. :(
Took me so much time with the slow testing (build pkg).

kewang1024 reviewed (commented):
Thanks @spershin for helping improve the verification of array(double)

spershin commented on presto-verifier/src/main/java/com/facebook/presto/verifier/framework/VerifierConfig.java:
```diff
@@ -248,6 +249,19 @@ public VerifierConfig setAbsoluteErrorMargin(double absoluteErrorMargin)
+        return arrayFloatingPointErrorMargin;
+    }
+
+    @ConfigDescription("When set to true, arrays of floating point numbers are validated like floating point columns, using error margins. False by default.")
+    @Config("array-floating-point-error-margin")
+    public VerifierConfig setArrayFloatingPointErrorMargin(boolean arrayFloatingPointErrorMargin)
+    {
+        this.arrayFloatingPointErrorMargin = arrayFloatingPointErrorMargin;
+        return this;
+    }
+
```
Yes, for rollout, I plan to set it to 'false' by default and then enable and use for a while and then deprecate.
Now it is true, so I can test using the builds.

kewang1024 commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/FloatingPointColumnChecksum.java:
```diff
@@ -86,4 +101,63 @@ public String toString()
+        }
+    }
+
+    public static ValidateResult validate(
```
Even changed to static function, I think it makes more sense to still put this function in [FloatingPointColumnValidator.java](https://github.com/prestodb/presto/pull/22143/files#diff-69d71b377216251d0931b7431965855d0a64922052e1167ad998c848f0b2cf5e)

FloatingPointColumnChecksum is only the data structure, FloatingPointColumnValidator will handle the validation logic

spershin replied:
@kewang1024 

Initially I kept it in the FloatingPointColumnValidator, but then I thought that this logic is better to be incapsulated in the structure that keeps the data. It already implements equals(), toString().
Make sense to put validate() here too, imho.

Let me know if you really prefer it back to the validator class.

rschlussel replied:
I think either all validation in should be in the respective checksum classes or none of them(also, might fit better in the validate class if it were a non-static method like equals e.g. `floatingPointChecksum.validate(FloatingPointChecksum other)`). As things stand it seems a bit out of place, and I agree with @kewang1024 that it would be better in the validate class.

spershin commented:
The last update:
1. Move ArrayColumnValidator.validate to be above private methods as per suggestion.
2. Remove unused ChecksumResult.hasChecksum()
3. Refactor ArrayColumnChecksum to contain FloatingPointColumnChecksum. Later we can use the same trick for Map keys and values.

spershin commented:
@kewang1024 , @rschlussel 
This change is close to the final change.
Code review is appreciated.
The previous change works successfully on the 10 NGA queries batch.
Will test on the real workload the final change (after comments are addressed) before merging.

kewang1024 reviewed (commented):
Thanks again! Overall looks good, some NIT

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:
```diff
@@ -25,12 +32,43 @@ public class ArrayColumnChecksum
 
-    public ArrayColumnChecksum(@Nullable Object checksum, @Nullable Object cardinalityChecksum, long cardinalitySum)
+    // Constructor for array(non- floating point) types.
+    public ArrayColumnChecksum(
```
it would be better to have one constructor that does the assignment and the other constructors call that one and pass in whatever args are needed.

spershin replied:
I did this for FloatingPointColumnChecksum, but for ArrayColumnChecksum I can only do this for empty ctor, as we need to initialize one member as optional in some cases.

kewang1024 replied:
I think providing two constructors here could be more straightforward

```
public ArrayColumnChecksum(@Nullable Object checksum, @Nullable Object cardinalityChecksum, long cardinalitySum)
{
this(checksum, cardinalityChecksum, cardinalitySum, Optional.empty());
}
```

```
public ArrayColumnChecksum(
@Nullable Object checksum,
@Nullable Object cardinalityChecksum,
long cardinalitySum,
Optional<FloatingPointColumnChecksum> floatingPointChecksum
)
{
// assignment
}
```

spershin replied:
@kewang1024 
> I think providing two constructors here could be more straightforward

```
Hmm, not sure if looked at the new code or the old one?
We have 3 ctors now:

    // Constructor for array(non-floating point) types.
    public ArrayColumnChecksum(
            @Nullable Object checksum,
            @Nullable Object cardinalityChecksum,
            long cardinalitySum)

    // Constructor for array(floating point) types.
    public ArrayColumnChecksum(
            @Nullable Object sum,
            long nanCount,
            long positiveInfinityCount,
            long negativeInfinityCount,
            @Nullable Object cardinalityChecksum,
            long cardinalitySum)

    // Constructor to handle cases of array(floating point) when the result table is empty (sum aggregations returns null).
    public ArrayColumnChecksum()

```

kewang1024 replied:
yeah, I saw that
combined with @rschlussel's suggestion, I think changing the constructors to the above two constructors can be more starightforward?

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:
```diff
@@ -63,19 +101,44 @@ public boolean equals(Object obj)
+        }
+    }
+
+    public static FloatingPointColumnChecksum.ValidateResult validateFloatingPoint(
```
it's weird to me that this logic is over here, but the rest of the validation logic is in ArrayColumnValidator.  I would do one or the other.

spershin replied:
Moved validate() logic to Validate classes.

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:
```diff
@@ -63,19 +101,44 @@ public boolean equals(Object obj)
+        // Check the non floating point members first.
+        if (!Objects.equals(controlChecksum.getCardinalityChecksum(), testChecksum.getCardinalityChecksum()) ||
+                !Objects.equals(controlChecksum.getCardinalitySum(), testChecksum.getCardinalitySum())) {
+            return new FloatingPointColumnChecksum.ValidateResult(false, Optional.of("cardinality mismatch"));
```
Do we need this new class? Can we use ColumnMatchResult instead?

spershin replied:
ColumnMatchResult is a templated class and we must return different instantiations of it from ArrayColumnChecksum and FloatingPointColumnChecksum.

I decide to have that small class to carry actual validation result, except all the columns info, which will be provided by concrete validators.

kewang1024 replied:
The `ValidateResult` still feel not needed, I don't see particular bad effect of using `ColumnMatchResult`

spershin replied:
Ok, I guess, we can get rid of it.

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -16,36 +16,87 @@
 {
+    private final double relativeErrorMargin;
+    private final double absoluteErrorMargin;
+    private final boolean arrayFloatingPointErrorMargin;
```
I find this name unclear.  Maybe something like "useErrorMarginForFloatArrays"?

spershin replied:
Replaced by useErrorMarginForFloatingPointArrays and "use-error-margin-for-floating-point-arrays".

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -16,36 +16,87 @@
+    @Inject
+    public ArrayColumnValidator(VerifierConfig config)
+    {
+        this.relativeErrorMargin = config.getRelativeErrorMargin();
```
I wonder if we should inject FloatingPointColumnValidator instead, and then can just call floatingPointColumnValidator.validate() and leave these configs as implementation details of the floatingpointcolumnvalidator.

spershin replied:
@rschlussel 
Do you mean make FloatingPointColumnValidator a member of ArrayColumnValidator and pass the former to the constructor of the latter?
We can do this.

spershin commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/FloatingPointColumnChecksum.java:
```diff
@@ -37,6 +42,16 @@ public FloatingPointColumnChecksum(@Nullable Object sum, long nanCount, long pos
         this.rowCount = rowCount;
     }
 
+    // Constructor to handle cases when the result table is empty (sum aggregations returns null).
```
Actually, I would prefer to leave it.
We use it, check the updated changes.

spershin commented:
@kewang1024 @rschlussel 

Thanks for the review.
I have addressed the comments so far.

spershin commented:
Addressed ArrayColumnChecksum ctor comments.

rschlussel commented on presto-verifier/src/test/java/com/facebook/presto/verifier/resolver/TestStructuredColumnMismatchResolver.java:
```diff
@@ -42,10 +44,10 @@ public TestStructuredColumnMismatchResolver()
     @Test
     public void testResolveArray()
```
Can you update the StructuredColumnsMismatchResolver and then add a test that we don't resolve column mismatches for float arrays if we're using error margin validation.

rschlussel commented on presto-verifier/src/test/java/com/facebook/presto/verifier/checksum/TestChecksumValidator.java:
```diff
@@ -245,6 +275,75 @@ public void testFloatingPoint()
     }
 
+    @Test
+    public void testFloatingPointArray()
```
can you add a test for the empty table case too?

spershin replied:
Yes, I was going to, but hit that snag with nulls in the Map.
I can see if I can use that HashMap to generate the nulls.

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:
```diff
@@ -63,19 +78,29 @@ public boolean equals(Object obj)
     public String toString()
     {
-        return format("checksum: %s, cardinality_checksum: %s, cardinality_sum: %s", checksum, cardinalityChecksum, cardinalitySum);
+        if (!floatingPointChecksum.isPresent()) {
```
nit: instead of repeating everything for constructing the string, you can use MoreObject.toStringHelper to construct a nicely printed string with all the fields. Something like:
```
MoreObjects.ToStringHelper toStringHelper = MoreObjects.toStringHelper(ArrayColumnchecksum)
    .add("cardinality_checksum", cardinalityChecksum)
    .add("cardinality_sum", cardinalitySum)

if(floatingPointChecksum.isPresent() {
    toStringHelper.add("floating_point_checksum", floatingPointchecksum.get());
} 
else {
    toStringHelper.add("checksum", checksum);
}

return toStringHelper.toString();
```

rschlussel replied:
not sure if it's important that this toString matches the toStrings of the other checksums, so I'll add the caveat that the generated string will look slightly different than what you have, see the docs for example output https://guava.dev/releases/19.0/api/docs/com/google/common/base/MoreObjects.html#toStringHelper(java.lang.Object)

spershin replied:
@rschlussel 
I don't know. I don't feel that this change worth the trouble at the moment.
You are saying "to construct a nicely printed string with all the fields" - are they going to be more nicely printed than with the current code? How?

Maybe we can consider a small refactor followup to move all Checksums to such format?

rschlussel replied:
not necessarily nicer than the current code. just helps you with not duplicating for the floating point vs. non-floating point case.  It would look something like `ArrayColumnChecksum{cardinality_checksum=123, cardinality_sum=345, ...}`. anyway, fine to leave it as is.

rschlussel commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnChecksum.java:
```diff
@@ -63,19 +78,29 @@ public boolean equals(Object obj)
+        else {
+            return format(
+                    "%s, cardinality_checksum: %s, cardinality_sum: %s",
+                    floatingPointChecksum.get().toString(),
```
fyi, you don't need to call toString() here. The format function will do it if you pass in the object.

rschlussel reviewed (commented):
Thanks for the update. A few comments, but generally looks good!

spershin commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -16,36 +16,85 @@
+    @Inject
+    public ArrayColumnValidator(VerifierConfig config)
+    {
+        this.floatingPointValidator = new FloatingPointColumnValidator(config);
```
@rschlussel 

I don't understand this. :)
Where do we take a floatingPointColumnValidator from to pass it into the ctor?
Are these Validator classes some kind of singletons and are being reused for multiple columns?
Thanks!

spershin commented on presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/ArrayColumnValidator.java:
```diff
@@ -55,10 +104,31 @@ public List<SingleColumn> generateChecksumColumns(Column column)
     {
-        ArrayColumnChecksum controlChecksum = toColumnChecksum(column, controlResult);
-        ArrayColumnChecksum testChecksum = toColumnChecksum(column, testResult);
+        checkArgument(
```
I'm not sure about it, honestly - repeating what FloatingPointColumnValidator is doing.

spershin replied:
Yes, it looks like a sanity check to me.
I think we can remove it from ArrayColumnValidator and move it in FloatingPointColumnValidator from overridden validate() to the public validate(), which is called from the ArrayColumnValidator.
